### PR TITLE
Modify (Large/Source)Text.ParseLineStart to specify an initial capacity for line count

### PR DESCRIPTION
**Pull request by ToddGrun**
Noticed this in a Find All References profile, the LargeText.ParseLineStarts was showing as 6.2% of allocations and SourceText.ParseLineStarts was 0.4%.

With this change, these reduced to 4.7% and 0.2% respectively.

*** before ***
image
and
image

*** after ***
image
and
image

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/Core/Portable/Text/LargeText.cs [hunk 1]
```diff
@@ -238,7 +238,8 @@ private SegmentedList<int> ParseLineStarts()
             var position = 0;
             var index = 0;
             var lastCr = -1;
-            var list = new SegmentedList<int>();
+            // Initial line capacity estimated at 64 chars / line
+            var list = new SegmentedList<int>(Length / 64);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Text/LargeText.cs:242]: you might want to mention this was based on looking at roslyn files and coming up with a reasonable balance.

Comment (ToddGrun) [reply src/Compilers/Core/Portable/Text/LargeText.cs:242]: Done!

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Text/LargeText.cs:242]: You say that.  And yet...

Comment (ToddGrun) [reply src/Compilers/Core/Portable/Text/LargeText.cs:243]: Guess I was in too big of a hurry to get home to push the button

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler ptal

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler -- ptal

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler -- ptal

File: src/Compilers/Core/Portable/Text/LargeText.cs [hunk 2]
```diff
@@ -238,7 +238,9 @@ private SegmentedList<int> ParseLineStarts()
             var position = 0;
             var index = 0;
             var lastCr = -1;
-            var list = new SegmentedList<int>();
+            // Initial line capacity estimated at 64 chars / line. This values was obtained by
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/Text/LargeText.cs:241]:
```suggestion
            // Initial line capacity estimated at 64 chars / line. This value was obtained by
```

File: src/Compilers/Core/Portable/Text/SourceText.cs [hunk 3]
```diff
@@ -1049,7 +1049,9 @@ private SegmentedList<int> ParseLineStarts()
                 return [0];
             }
 
-            var lineStarts = new SegmentedList<int>()
+            // Initial line capacity estimated at 64 chars / line. This values was obtained by
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/Text/SourceText.cs:1052]:
```suggestion
            // Initial line capacity estimated at 64 chars / line. This value was obtained by
```

**Review by jjonescz — APPROVED** [review]

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler -- ptal

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/Core/Portable/Text/LargeText.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/Text/LargeText.cs:243]: Just want to add that I really appreciate context like this for magic numbers.

**Pull request merged**